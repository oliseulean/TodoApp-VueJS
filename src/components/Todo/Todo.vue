<template>
  <div class="container">
    <header class="header">
      <input
        type="text"
        class="box"
        placeholder="Add item/s"
        autofocus
        @keyup.enter="addItem"
        v-model="newItem"
      />
      <button class="newItem" @click="addItem">+</button>
    </header>
    <div class="items">
      <ul class="todo-list">
        <li
          v-for="(item, index) in itemsFiltered"
          :class="{ completed: item.completed }"
          :key="index"
        >
          <div class="view">
            <input
              type="checkbox"
              :checked="item.completed"
              @change="completeItem(item)"
            />
            <label>{{ item.name }}</label>
            <div class="delete" @click="removeItem(item)">X</div>
          </div>
        </li>
      </ul>
    </div>
    <footer class="footer" v-if="items.length > 0">
      <div class="filters">
        <li
          @click="
            changeFilter('all');
            makeActive(1);
          "
          :class="{ active: activeElement === 1 }"
        >
          All
        </li>
        <li
          @click="
            changeFilter('active');
            makeActive(2);
          "
          :class="{ active: activeElement === 2 }"
        >
          Active
        </li>
        <li
          @click="
            changeFilter('completed');
            makeActive(3);
          "
          :class="{ active: activeElement === 3 }"
        >
          Completed
        </li>
        <li @click="clearAll">
          Clear All
        </li>
        <li
          v-if="items.length > showClearCompletedButoon"
          @click="clearCompleted"
        >
          Clear completed
        </li>
      </div>
    </footer>
    <div class="nrOfItems" v-if="items.length > 0">
      <span>{{ this.itemsLeft }}</span>
      {{ showItemsLeft }}
    </div>
  </div>
</template>

<script src="./index.js" />
<style>
@import "./index.css";
</style>
