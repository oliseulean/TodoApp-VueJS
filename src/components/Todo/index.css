body {
	background-color: #F5F5F5;
	color: #2c3e50;
}

.container {
	display: flex;
	margin: auto;
	flex-direction: column;
	background-color: white;
	box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.2), 0 25px 50px 0 rgba(0, 0, 0, 0.1);
}

.header {
	display: flex;
	margin: 20px;
	height: 50px;
}

.header ::placeholder {
	font-style: italic;
	opacity: 0.5;
	font-size: 16px;
}

.box {
	border: none;
	width: 90%;
	padding: 16px 60px 16px 60px;
	border: none;
	background: #efefef;
	transition: background .3s ease;
}

.box:hover {
	border: 1px solid black;
}

.newItem {
	background: #41B883;
	color: white;
	border: none;
	font-size: 40px;
	width: 10%;
}

.todo-list {
	margin: 0;
	padding: 0;
	list-style-type: none;
}

.new-todo-button {
	display: flex;
	flex-direction: column;
}

input[type="text"] {
	outline: none;
}

.todo-list li {
	border-bottom: 1px solid #ededed;
	padding: 10px;
	height: 50px;
}

.view {
	position: relative;
	height: 100%;
}

.view label {
	font-size: 16px;
	padding-left: 20px;
}

.view:hover .delete {
	visibility: visible;
}

.delete {
	position: absolute;
	right: 5px;
	top: 3px;
	height: 20px;
	visibility: hidden;
	color: red;
	cursor: pointer;
}

.footer {
	padding: 10px 15px;
	color: #8D7777;
	height: 50px;
	display: block;
	text-align: center;
}

.filters {
	display: block;
	right: 0;
	left: 0;
}

.filters li {
	list-style-type: none;
	padding-left: 5px;
	padding-right: 5px;
}

.completed {
	text-decoration-line: line-through;
	color: gray;
}

.active {
	border-bottom: 1px #41B883 solid;
	font-weight: bold;
	color: #41B883;
}

.items {
	padding: 8px 0;
}

.items ul.todo-list li:hover {
	background: #f7fcfa;
}

input[type=checkbox] {
	cursor: pointer;
	filter: hue-rotate(280deg);
	background-color: #41B883;
}

.nrOfItems {
	display: flex;
	text-align: center;
	padding: 10px;
	background-color: #41B883;
}

.filters {
	width: 100%;
	display: flex;
	font-size: 13px;
	cursor: pointer;
}

.nrOfItems span {
	padding-right: 2px;
	font-weight: bold;
}