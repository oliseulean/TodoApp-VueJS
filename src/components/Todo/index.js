import createUID from "create-unique-id";

export default {
  name: "App",
  data() {
    return {
      items: [],
      newItem: "",
      itemsLeft: null,
      filter: "all",
      activeElement: 1
    };
  },
  created() {
    this.items = JSON.parse(localStorage.getItem("list") || "[]");
    this.itemsLeft = this.itemsFiltered.length;
  },
  computed: {
    itemsFiltered() {
      if (this.filter === "all") {
        return this.items;
      } else if (this.filter === "active") {
        return this.items.filter(item => !item.completed);
      } else if (this.filter === "completed") {
        return this.items.filter(item => item.completed);
      }
    },
    showClearCompletedButoon() {
      const getItems = this.items.filter(item => !item.completed);
      const itemsLength = getItems.length;
      return itemsLength;
    },
    showItemsLeft() {
      return this.itemsLeft === 0 || this.itemsLeft === 1
        ? `item left`
        : `items left`;
    }
  },
  methods: {
    addItem() {
      if (this.newItem.trim() != "") {
        this.items.unshift({
          id: createUID(10),
          completed: false,
          name: this.newItem
        });
        this.newItem = "";
        localStorage.setItem("list", JSON.stringify(this.items));
        this.itemsLeft = this.itemsFiltered.length;
      }
    },
    removeItem(item) {
      const itemIndex = this.items.indexOf(item);
      this.items.splice(itemIndex, 1);
      localStorage.setItem("list", JSON.stringify(this.items));
      this.itemsLeft = this.itemsFiltered.length;
    },
    getList() {
      const listString = localStorage.getItem("list");
      if (!listString) {
        return [];
      }
      return JSON.parse(listString);
    },
    completeItem(item) {
      item.completed = !item.completed;
      localStorage.setItem("list", JSON.stringify(this.items));
    },
    clearCompleted() {
      let list = this.getList();
      let listOfActiveItems = list.filter(key => {
        return !key.completed;
      });
      localStorage.setItem("list", JSON.stringify(listOfActiveItems));
      this.items = listOfActiveItems;
      this.itemsLeft = this.itemsFiltered.length;
    },
    changeFilter(filter) {
      this.filter = filter;
      this.itemsLeft = this.itemsFiltered.length;
    },
    clearAll() {
      this.items = [];
      this.itemsLeft = this.itemsFiltered.length;
      localStorage.setItem("list", JSON.stringify(this.items));
    },
    makeActive(element) {
      this.activeElement = element;
    }
  }
};
