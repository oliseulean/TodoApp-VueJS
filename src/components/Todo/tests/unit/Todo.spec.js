import { shallowMount } from "@vue/test-utils";
import Todo from "@/components/Todo/Todo";

const findById = (id, list) => list.find(item => item.id === id);
const removeTodo = (list, todoId) => {
  const todoIndex = list.findIndex(item => item.id === todoId);
  return [...list.slice(0, todoIndex), ...list.slice(todoIndex + 1)];
};

// jest.mock("uuid/v4");

describe("Todo component", () => {
  test("Todo is a component", () => {
    const wrapper = shallowMount(Todo);
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("Render html correctly", () => {
    const wrapper = shallowMount(Todo);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Add button appears", () => {
    const addButton = ".newItem";
    const wrapper = shallowMount(Todo);
    expect(wrapper.contains(addButton)).toBe(true);
  });
});

describe("Adding TodoItems", () => {
  const wrapper = shallowMount(Todo);

  async function addTodo(todoText) {
    const inputButton = ".box";
    const addButton = ".newItem";
    wrapper.find(inputButton).setValue(todoText);
    await wrapper.find(addButton).trigger("click");
  }

  test("allows for adding one todo item", async () => {
    await addTodo("Test");
    const elements = ".todo-list";
    expect(wrapper.find(elements).text()).toContain("Test");
  });
  test("empties the input field when todo has been added", async () => {
    const inputButton = ".box";
    await addTodo("This is not important");
    const findInputButton = wrapper.find(inputButton);
    expect(findInputButton.element.value).toEqual("");
  });

  describe("delete a todo", () => {
    test("should have todo removed", async () => {
      const deleteItem = jest.spyOn(Todo.methods, "removeItem");
      const items = [
        { id: 1, description: "go to the gym", isComplete: false }
      ];
      const wrapper = shallowMount(Todo, items);
      await wrapper.find(".delete").trigger("click");
      expect(deleteItem).toHaveBeenCalledTimes(1);
    });
    test("removeTodo by id should not mutate", () => {
      const startTodos = [
        { id: 1, description: "go to the gym", isComplete: false },
        { id: 2, description: "get food", isComplete: false },
        { id: 3, description: "clean the room", isComplete: false }
      ];

      const expected = [
        { id: 1, description: "go to the gym", isComplete: false },
        { id: 3, description: "clean the room", isComplete: false }
      ];

      const result = removeTodo(startTodos, 2);

      expect(result).not.toBe(expected);
    });
  });

  test("findById should return the expected item from array", () => {
    const startTodos = [
      { id: 1, description: "go to the gym", isComplete: false },
      { id: 2, description: "get food", isComplete: false },
      { id: 3, description: "clean the room", isComplete: false }
    ];

    const expected = { id: 2, description: "get food", isComplete: false };

    const result = findById(2, startTodos);

    expect(result).toEqual(expected);
  });
});
