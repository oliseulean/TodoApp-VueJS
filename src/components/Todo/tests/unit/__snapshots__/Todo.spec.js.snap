// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo component Render html correctly 1`] = `
<div class="container">
  <header class="header"><input type="text" placeholder="Add item/s" autofocus="autofocus" class="box"> <button class="newItem">+</button></header>
  <div class="items">
    <ul class="todo-list"></ul>
  </div>
  <!---->
  <!---->
</div>
`;
